nab/detectors/htmcore/htmcore_detector.py

```diff
@@ -332,6 +334,12 @@ def modelRun(self, ts, val):
           pass
 
       if PANDA_VIS_ENABLED:
+          # activateDendrites calculates active segments
```

Comment:
this if-else starts to be complicated. 
- what used to be `tm.compute()` https://github.com/htm-community/NAB/pull/36/files#diff-d9640827e1a8cfdf99cc729ed05d01ecR291 
- became (understendable) split to `activate dendrites & activate cells` : https://github.com/htm-community/NAB/pull/36/files#diff-d9640827e1a8cfdf99cc729ed05d01ecL281-L286 
- now the split to the 2nd (3rd) if-else branch is confusing. 

Note, now reviewing it, without more deeply thinking about it and verifying, I'd stay on the safe side and say that the order should be as in `TemporalMemory.cpp compute()` - so 1st dendrites, 2nd cells. That will give you valid results for both.

Reply:
I agree that it is confusing, certainly much more that it was.

But the problem is, if we leave here this

> became (understendable) split to activate dendrites & activate cells : https://github.com/htm-community/NAB/pull/36/files#diff-d9640827e1a8cfdf99cc729ed05d01ecL281-L286

Then call for example self.tm_info.addData( self.tm.getActiveCells().flatten() )
then some pointer leakages can occur - it happened to me, i was hunting that bug and i figured out, that htm.core is not internally made to be able to do this. Note: i was using external distal inputs, i saw that the externalDistalInputs are literally added to the vector of ActiveCells and that is messing things up (actually it is then SDR of size let's say 65535 and the vector contains indexes like 65536 and more) so we can't read getActiveCells between activaeDendrites and activateCells

Reply:
If you think that would be better to create htmcore_detector_pandaVis.py instead to incorporating it to this script that is used as benchmark/optimizer i would not be against. I didn't know at the beggining that it will gets more complicated like this

Reply:
> Then call for example self.tm_info.addData( self.tm.getActiveCells().flatten() )
then some pointer leakages can occur - it happened to me, i was hunting that bug and i figured out, that htm.core is not internally made to be able to do this. 

can you make a reproducer for that? Sounds like something that should ideally be ironed out in the lib itself. Unless you're doing something that's marked as undefined/incorrect behavior (see c++'s doc for the relevant methods in TM for this please) 

>  Note: i was using external distal inputs, i saw that the externalDistalInputs are literally added to the vector of ActiveCells and that is messing things up (actually it is then SDR of size let's say 65535 and the vector contains indexes like 65536 and more) 

hmm, yes, this is sort of hacked-in code. But as I recall the code is correct, "external inputs" are indeed _external_, the cells' indices are out-of those that the TM/connections use. And if you use external inputs, you should expect that. I haven't worked with that code lately, so I'd certainly welcome feedback and discussion. We might try to find better ways to handle that. 

> so we can't read getActiveCells between activaeDendrites and activateCells 

can't we? 
From the `activateDendrites` doc

> * Calculate dendrite segment activity, using the current active cells.  Call
* this method before calling getPredictiveCells, getActiveSegments, or
* getMatchingSegments.  In each time step, only the first call to this
* method has an effect, subsequent calls assume that the prior results are
* still valid. [...]

All in all, I hope we can improve the API,doc, usage of the external dendrites with this.

Reply:
> If you think that would be better to create htmcore_detector_pandaVis.py instead to incorporating it to this script that is used as benchmark/optimizer i would not be against.  

let's see how it goes. I certainly don't, as I asked you if you could merge the two. On the other hand, things are getting to be more complicated than when we started. But also, once done, the detector code wouldn't be actively used (changed) by the users, merely changing the constants and calling it.